cmd/util/cmd/epochs/cmd/recover.go

```diff
@@ -83,103 +72,142 @@ func generateRecoverEpochTxArgs(cmd *cobra.Command, args []string) {
+		log.Fatal().Err(err).Msg("failed to read full internal node infos")
 	}
 
-	log.Info().Msg("collecting partner network and staking keys")
-	partnerNodes := common.ReadPartnerNodeInfos(log, flagPartnerWeights, flagPartnerNodeInfoDir)
+	internalNodesMap := make(map[flow.Identifier]struct{})
+	for _, node := range internalNodes {
+		if !ids.Exists(node.Identity()) {
+			log.Fatal().Msg(fmt.Sprintf("node ID found in internal node infos missing from protocol snapshot identities: %s", node.NodeID))
+		}
+		internalNodesMap[node.NodeID] = struct{}{}
+	}
 	log.Info().Msg("")
 
-	log.Info().Msg("generating internal private networking and staking keys")
-	internalNodes := common.ReadInternalNodeInfos(log, flagInternalNodePrivInfoDir, flagConfig)
-	log.Info().Msg("")
+	for _, collector := range collectors {
+		if _, ok := internalNodesMap[collector.NodeID]; ok {
+			internalCollectors = append(internalCollectors, collector)
+		} else {
+			partnerCollectors = append(partnerCollectors, collector)
+		}
+	}
+
+	currentEpochDKG, err := epoch.DKG()
+	if err != nil {
+		log.Fatal().Err(err).Msg("failed to get DKG for current epoch")
+	}
 
 	log.Info().Msg("computing collection node clusters")
-	_, clusters, err := common.ConstructClusterAssignment(log, partnerNodes, internalNodes, flagCollectionClusters)
+
+	assignments, clusters, err := common.ConstructClusterAssignment(log, partnerCollectors, internalCollectors, flagCollectionClusters)
 	if err != nil {
 		log.Fatal().Err(err).Msg("unable to generate cluster assignment")
 	}
 	log.Info().Msg("")
 
-	epochCounter, err := epoch.Counter()
-	if err != nil {
-		log.Fatal().Err(err).Msg("unable to get epoch counter from current epoch")
-	}
 	log.Info().Msg("constructing root blocks for collection node clusters")
-	clusterBlocks := run.GenerateRootClusterBlocks(epochCounter, clusters)
+	clusterBlocks := run.GenerateRootClusterBlocks(flagEpochCounter, clusters)
 	log.Info().Msg("")
 
 	log.Info().Msg("constructing root QCs for collection node clusters")
 	clusterQCs := common.ConstructRootQCsForClusters(log, clusters, internalNodes, clusterBlocks)
-	fmt.Sprintf("", clusterQCs)
 	log.Info().Msg("")
 
-	randomSource, err := epoch.RandomSource()
-	if err != nil {
-		log.Fatal().Err(err).Msg("failed to get random source for current epoch")
-	}
-	randomSourceCdc, err := cadence.NewString(string(randomSource))
-	if err != nil {
-		log.Fatal().Err(err).Msg("failed to get random source cadence string")
-	}
-
 	dkgPubKeys := make([]cadence.Value, 0)
 	nodeIds := make([]cadence.Value, 0)
-	ids.Map(func(skeleton flow.IdentitySkeleton) flow.IdentitySkeleton {
-		if skeleton.GetRole() == flow.RoleConsensus {
-			dkgPubKey, keyShareErr := currentEpochDKG.KeyShare(skeleton.GetNodeID())
+
+	dkgGroupKeyCdc, cdcErr := cadence.NewString(currentEpochDKG.GroupKey().String())
+	if cdcErr != nil {
+		log.Fatal().Err(cdcErr).Msg("failed to get dkg group key cadence string")
+	}
+	dkgPubKeys = append(dkgPubKeys, dkgGroupKeyCdc)
```

Comment:
In this implementation we are requiring that the consensus committee is identical in the last epoch and in the recovery epoch (no changes to the committee is allowed during EFM recovery process is allowed). We should document this constraint (I'll add some suggestions to the cmd documentation section as well).
```suggestion
	// NOTE: The RecoveryEpoch will re-use the last successful DKG output. This means that the consensus
	// committee in the RecoveryEpoch must be identical to the committee which participated in that DKG.
	dkgGroupKeyCdc, cdcErr := cadence.NewString(currentEpochDKG.GroupKey().String())
	if cdcErr != nil {
		log.Fatal().Err(cdcErr).Msg("failed to get dkg group key cadence string")
	}
	dkgPubKeys = append(dkgPubKeys, dkgGroupKeyCdc)
```